### ISSUE TITLE
max_tokens_per_request policy only caps input, not output; default output limit is 96 tokens; no streaming support

### ISSUE THREAD

**Issue by shivakumardistproc** [Ephapsys/ephapsys-samples#3]
Summary                                                                 

  Two issues with language model inference in the SDK.                                                                                                                    
  
  1. No streaming output support                                                                                                                                          
                                                                                                                                                                          
  All inference in _run_language() is synchronous — model.generate() blocks until the full output is produced (agent.py:3275). There is no streaming path (e.g. HuggingFace TextIteratorStreamer) available.
                                                                                                                                                                          
  For interactive/chat use cases, streaming would significantly improve UX.  

  2. max_tokens_per_request does not limit output tokens                                                                                                                  
                                                                          
  The "Max Tokens / Request" field in the Agent Template UI (default 4096) maps to the max_tokens_per_request policy. However, in the SDK this value only controls input   truncation, not output generation:
                                                                                                                                                                          
  ### agent.py:2521 — only sets input cap                                   
  cap = int(policy.get("max_tokens_per_request") or DEFAULT_MAX_TOKEN_LENGTH)                                                                                             
  self._max_tokens_cap = max(128, min(cap, DEFAULT_MAX_TOKEN_LENGTH * 4))                                                                                                 
                                                                                                                                                                          
  ### agent.py:769 — used only for input                                                                                                                                    
  if phase == "input" and len(toks) > self._max_tokens_cap:                                                                                                               
      value = " ".join(toks[:self._max_tokens_cap])                                                                                                                       
   
  The actual output token limit comes from a separate max_new_tokens config path which falls through to a hardcoded default of 96:                                        
                                                                          
  ### agent.py:3234-3235                                                                                                                                                    
  max_new_tokens = int(                                                   
      generation_cfg.get("max_new_tokens", model_cfg.get("max_new_tokens", os.getenv("AOC_MAX_NEW_TOKENS", "96")))                                                        
  )                                                                                                                                                                       
                                                                                                                                                                          
  Expected: The max_tokens_per_request policy should either:                                                                                                              
  - Also inform the output generation limit, or                           
  - The UI should clearly distinguish between input cap and output cap, with a separate configurable field for max_new_tokens                                             
                                                                                                                             
  Workaround: Set AOC_MAX_NEW_TOKENS=1024 as an environment variable.                                                                                                     
                                                                                                                                                                          
<img width="1920" height="1080" alt="Image" src="https://github.com/user-attachments/assets/bfbc6d0c-d796-4169-a3d2-1558389fa022" />
                                                                        
  The "Max Tokens / Request" field in the Create Agent Template UI showing 4096 — this value does not control output length as one might expect.

**Comment by izzowane** [comment]
Both addressed in the SDK (`ephapsys-sdk@8d1e5c90`).

## 1. Output token cap now honors `max_tokens_per_request`

`_run_language` no longer falls through to the hardcoded 96. New `max_new_tokens` precedence:

```
explicit generation cfg / model cfg / AOC_MAX_NEW_TOKENS env
  > max_tokens_per_request policy cap (self._max_tokens_cap)
  > 512 fallback
```

So the Agent Template "Max Tokens / Request" field now actually governs response length. The `AOC_MAX_NEW_TOKENS` env still works as an explicit override; the floor fallback is 512 instead of the surprising 96.

## 2. Streaming output

New `TrustedAgent.run_stream(input, model_kind="language")` yields decoded text chunks as they generate (HF `TextIteratorStreamer`, generation on a background thread):

```python
for chunk in agent.run_stream("Tell me a story", model_kind="language"):
    print(chunk, end="", flush=True)
```

Same fail-closed preflight as `run()` (enabled / personalized / not-revoked, per-model selective-inference status, input policies). Output policies + guardrails are enforced on the full accumulated text when the stream completes — callers needing pre-emptive output redaction should use the non-streaming `run()`. GGUF runtimes degrade to a single synchronous chunk.

Internals: model-load and generate-kwargs setup were extracted into `_ensure_language_model_loaded()` and `_prepare_language_inputs()` so the sync and streaming paths share identical setup (and the cap precedence lives in one place).

## Local testing

Added `tests/test_language_streaming_and_cap.py` — runs fully offline (tiny model, no AOC/auth):

```
cd ephapsys-sdk
pip install transformers torch        # or: pip install -e .
python tests/test_language_streaming_and_cap.py
```

Verified locally — all pass:
- policy cap drives `max_new_tokens=64`
- fallback is 512, not 96
- explicit generation cfg overrides the policy cap
- streamed output (17 chunks) concatenates exactly to the sync output under greedy decode

Note: this lands in source on `main`; it'll reach agents after the next SDK PyPI publish + version bump.

**Comment by izzowane** [comment]
Shipped in **SDK 0.2.93** (now on PyPI: https://pypi.org/project/ephapsys/0.2.93/).

Both reported items are fixed, plus a related determinism bug surfaced while writing the local test:

1. **Output token cap** — `max_new_tokens` now derives from the `max_tokens_per_request` policy when no explicit generation value is set (precedence: explicit generation/model cfg / `AOC_MAX_NEW_TOKENS` env > policy cap > 512 fallback). No more hardcoded 96; the "Max Tokens / Request" field governs response length.

2. **Streaming** — `TrustedAgent.run_stream(input, model_kind="language")` yields decoded chunks via HF `TextIteratorStreamer`, with the same fail-closed governance preflight as `run()`.

3. **Bonus fix — eval mode.** The SDK never called `model.eval()` after loading, so dropout stayed active and language inference was non-deterministic across calls (sync vs streamed greedy decodes diverged on identical input). Fixed in `_ensure_language_model_loaded` — decoding is now reproducible. This affected all prior versions.

Local test added: `ephapsys-sdk/tests/test_language_streaming_and_cap.py` (offline, tiny model) — verifies cap precedence and that streamed output matches sync byte-for-byte.

Note: deployed agents pick up 0.2.93 after the AOC frontend redeploys with `NEXT_PUBLIC_SDK_VERSION=0.2.93` (already committed).

Closing as resolved.
